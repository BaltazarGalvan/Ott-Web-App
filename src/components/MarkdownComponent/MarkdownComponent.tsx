import React from 'react';

const MARKDOWN_LINK_REGEX = /\[([^[]+)]\(((https?:\/\/|www\.)?[^)]+)\)/gi;
const MARKDOWN_ITALIC_REGEX = /(?:\*|_)(\S[\s\S]*?)(?:\*|_)/gi;
const MARKDOWN_STRONG_REGEX = /(?:\*{2}|_{2})(\S[\s\S]*?)(?:\*{2}|_{2})/gi;
const MARKDOWN_ITALIC_STRONG_REGEX = /(?:\*{3}|_{3})(\S[\s\S]*?)(?:\*{3}|_{3})/gi;
const MARKDOWN_HEADERS_REGEX = /^(#{1,6})(.*)$/gm;
const LINEBREAK_REGEX = /(?:\r\n|\r|\n)/g;

type Props = {
  markdownString: string;
};

const parseMarkdown = (value: string): string =>
  value
    .replace(/<[^>]+>/gm, '') // remove html should run first
    .replace(MARKDOWN_HEADERS_REGEX, (...[, header, title]) => {
      return '<h' + header.length + '>' + title.trim() + '</h' + header.length + '>';
    })
    .replace(MARKDOWN_ITALIC_STRONG_REGEX, (...[, word]) => `<strong><em>${word}</em></strong>`)
    .replace(MARKDOWN_STRONG_REGEX, (...[, word]) => `<strong>${word}</strong>`)
    .replace(MARKDOWN_ITALIC_REGEX, (...[, word]) => `<em>${word}</em>`)
    .replace(MARKDOWN_LINK_REGEX, (...[, word, link]) => {
      const target = /^(https?|www\.|\/\/)/.test(link) ? ' target="_blank"' : '';

      return `<a href="${link}"${target}>${word}</a>`;
    })
    .replace(LINEBREAK_REGEX, '<br />'); // linebreak formatter should run last

const MarkdownComponent: React.FC<Props> = ({ markdownString }: Props) => {
  return <div dangerouslySetInnerHTML={{ __html: parseMarkdown(markdownString) }} />;
};

export default MarkdownComponent;
