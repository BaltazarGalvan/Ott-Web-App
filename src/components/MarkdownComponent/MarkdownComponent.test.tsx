import React from 'react';
import { render } from '@testing-library/react';

import MarkdownComponent from './MarkdownComponent';

describe('<MarkdownComponent>', () => {
  test('renders and matches snapshot', () => {
    const markdownString = 'This is a test markdown string with a [testlink](https://www.example.com)';
    const { container } = render(<MarkdownComponent markdownString={markdownString} />);

    expect(container).toMatchSnapshot();
  });
  test('marks a link', () => {
    const markdownString = 'This is a test markdown string with a [testlink](https://www.example.com)';
    const { container } = render(<MarkdownComponent markdownString={markdownString} />);

    expect(container).toContainHTML('<a href');
  });
});
