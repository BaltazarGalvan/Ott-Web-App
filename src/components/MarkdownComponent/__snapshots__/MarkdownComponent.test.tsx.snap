// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MarkdownComponent> renders and matches snapshot 1`] = `
<div>
  <div>
    This is a test markdown string with a 
    <a
      href="https://www.example.com"
      target="_blank"
    >
      testlink
    </a>
  </div>
</div>
`;
